### PR TITLE
[WIP] Fixes #21964 - speed up host edit form rendering

### PR DESCRIPTION
**Pull request by justahero**
This PR addresses some of the performance issues seen on the host(s) page, in particular the edit page. The changes are based on #5095, a few more helpers were also refactored to only return the required properties used in the view templates, with the goal to avoid having to instantiate ActiveRecord models.

The overall improvements may not seem that big, mostly because the time spent is fairly distributed over many views. No single view or function seems to contribute in a significant way. Functions / plugins that execute HTTP requests with external services were not considered. Your mileage may vary, you may experience better performance overall. In our case we brought down the response time of `HostsController#edit` by about 10-20% (from around 5.8 to 4.2 seconds). In particular the view template listing Puppet classes should render faster.

The following changes were made to improve performance:

* refactor host related view templates, most notably `puppetclasses/_classes`, extract some variables to avoid re-calculations
* refactor helper functions in `ApplicationHelper` to return properties instead of instantiating AR models, especially for select boxes
* refactor some direct calls to ActiveRecord queries inside the views to return only the required columns, see e.g. `HttpProxy`, `ComputeProfile`

### CODE REVIEW

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -475,11 +475,25 @@ def accessible_resource_records(resource, order = :name)
...
+
+    list.reorder(order)
+  end
+
+  def accessible_resource_for_select(obj, resource, order: :name, association: resource, columns: [:id, :name])
+    columns = Array(columns)
+    accessible_resource(obj, resource, order, association: association).pluck(*columns)
   end
 
   def accessible_related_resource(obj, relation, order: :name, where: nil)
```
Comment (timogoebel) [inline app/helpers/application_helper.rb:494]: Nit: We should probably return `obj.none` (an empty active record result) instead of an empty array here.

File: app/views/puppetclasses/_classes.html.erb [hunk 2]
```diff
@@ -1,16 +1,23 @@
-<% Puppetclass.classes2hash(puppetclasses).sort.in_groups(2,nil) do |group| %>
+<% authorized = authorized_for(:controller => :host_editing, :action => :edit_classes) %>
```
Comment (timogoebel) [inline app/views/puppetclasses/_classes.html.erb]: Nit:

```suggestion
<% authorized = authorized_for(controller: :host_editing, action: :edit_classes) %>
```

**Comment by timogoebel** [comment]
ok to test

@tbrisker, @lzap, @ShimShtein: I hope you don't mind if I ask you for your opinion on this. 5.8 seconds load time for a form is just too much, that's why we invested quite a lot of time here to investigate this. I hope we can further improve this, but this is the first shot.

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -475,11 +475,25 @@ def accessible_resource_records(resource, order = :name)
...
+      current = obj.public_send(association)
+      # match the unscope and readonly values from the left query
+      list = list.or(klass.unscoped.unscope(list.unscope_values).readonly(list.readonly_value).where(:id => current.id)) if current.present?
+    end
+
+    list.reorder(order)
+  end
+
+  def accessible_resource_for_select(obj, resource, order: :name, association: resource, columns: [:id, :name])
+    columns = Array(columns)
```
Comment (kamils-iRonin) [inline app/helpers/application_helper.rb]: I think you can delete this line
```suggestion
```

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (justahero) [reply app/helpers/application_helper.rb:494]: Not sure about this, if records are fetched this function returns an array of tuples (arrays), not ActiveRecord models, therefore it cannot be chained with other query methods. Returning an empty array ensures the same return type.
I can test the code with `obj.none` and check that everything still works.

Comment (timogoebel) [reply app/helpers/application_helper.rb:494]: Are you sure? I thought it always returns an active record relation. Only the guard clause returns an empty array. And that's pretty inconsistent.

Comment (justahero) [reply app/helpers/application_helper.rb:494]: According to the [Active Record Querying](https://guides.rubyonrails.org/active_record_querying.html#pluck) documentation the method `pluck` returns an array.

File: app/views/compute_resources_vms/form/openstack/_base.html.erb [hunk 4]
```diff
@@ -8,7 +8,7 @@
                 {:inculde_blank => (compute_resource.zones.empty? || compute_resource.zones.size == 1) ? false : _("No preference") },
                 :label => _('Availability zone'), :label_size => "col-md-2" %>
 <div id='image_selection'>
-  <%= select_f f, :image_ref, images, :uuid, :name,
+  <%= select_f f, :image_ref, images, :first, :last,
```
Comment (kamils-iRonin) [inline app/views/compute_resources_vms/form/openstack/_base.html.erb]: since to the `select_f` method we always pass `:first` and `:last`, maybe these values should be set as default? or maybe we don't need these params and we can just pass only the array?

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (timogoebel) [reply app/helpers/application_helper.rb:494]: Yeah, but we don't use `pluck` in `accessible_related_resource` :-D

File: app/views/compute_resources_vms/form/openstack/_base.html.erb (hunk 4 above)
Comment (timogoebel) [reply app/views/compute_resources_vms/form/openstack/_base.html.erb]: That will probably break all plugins out there. I think we prefer staying API compatible.

File: app/controllers/hosts_controller.rb [hunk 5]
```diff
@@ -35,7 +35,7 @@ class HostsController < ApplicationController
                                           :runtime, :resources, :nics, :ipmi_boot, :console,
                                           :toggle_manage, :pxe_config, :disassociate, :build_errors]
 
-  before_action :taxonomy_scope, :only => [:new, :edit] + AJAX_REQUESTS
+  before_action :taxonomy_scope, :only => [:new] + AJAX_REQUESTS
```
Comment (tbrisker) [inline app/controllers/hosts_controller.rb]: Why was this removed? we need to maintain the taxonomy scope for the edited host.

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -475,11 +475,25 @@ def accessible_resource_records(resource, order = :name)
   end
 
   def accessible_resource(obj, resource, order = :name, association: resource)
-    list = accessible_resource_records(resource, order).to_a
+    klass = resource.to_s.classify.constantize
+
+    list = klass.unscoped.authorized
+
+    list = list.with_taxonomy_scope_override(@location, @organization) if klass.include? Taxonomix
```
Comment (tbrisker) [inline app/helpers/application_helper.rb:482]: how are these three lines different to the return value of  `accessible_resource_records`?

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -475,11 +475,25 @@ def accessible_resource_records(resource, order = :name)
...
+
     # we need to allow the current value even if it was filtered
-    current = obj.public_send(association) if obj.respond_to?(association)
-    list |= [current] if current.present?
-    list
+    if obj.respond_to?(association)
+      current = obj.public_send(association)
+      # match the unscope and readonly values from the left query
+      list = list.or(klass.unscoped.unscope(list.unscope_values).readonly(list.readonly_value).where(:id => current.id)) if current.present?
```
Comment (tbrisker) [inline app/helpers/application_helper.rb:488]: why do we need to match them? wouldn't `.or(where(id: current.id))`work just as well?

File: app/helpers/hosts_nic_helper.rb [hunk 8]
```diff
@@ -6,17 +6,22 @@ def suggest_new_link(form, field, link_class)
...
         :help_inline => :indicator,
         :'data-url' => freeip_subnets_path,
         :size => "col-md-8", :label_size => "col-md-3" }
     )
-    if accessible_resource(f.object, klass).any?
+    if subnets.any?
       array = options_for_select(
         [[]] +
-        accessible_resource(f.object, klass).map { |subnet| [subnet.to_label, subnet.id, {'data-suggest_new' => subnet.unused_ip.suggest_new?, 'data-vlan_id' => subnet.vlanid}]}, f.object.public_send(attr)
+        subnets.map { |subnet| [subnet[1], subnet[0], {'data-suggest_new' => false, 'data-vlan_id' => subnet[2]}]}, f.object.public_send(attr)
```
Comment (tbrisker) [inline app/helpers/hosts_nic_helper.rb]: `subnet.to_label != subnet.name`

File: app/views/common/os_selection/_architecture.html.erb [hunk 9]
```diff
@@ -1,5 +1,5 @@
 <%= fields_for item do |f| %>
-  <%= select_f f, :operatingsystem_id, arch_oss, :id, :to_label,
+  <%= select_f f, :operatingsystem_id, arch_oss, :first, :last,
```
Comment (tbrisker) [inline app/views/common/os_selection/_architecture.html.erb]: `os.to_label != os.name`

File: app/views/hostgroups/_form.html.erb [hunk 10]
```diff
@@ -49,14 +49,14 @@
         <%= select_f f, :domain_id, accessible_resource(@hostgroup, :domain), :id, :to_label, {:include_blank => blank_or_inherit_f(f, :domain)},
           {:help_inline => :indicator, :label => _("Domain"),
            :onchange => 'interface_domain_selected(this);', :'data-url' => method_path('domain_selected') } %>
-        <%= select_f f, :subnet_id, domain_subnets(:"Subnet::Ipv4"), :id, :to_label,
+        <%= select_f f, :subnet_id, domain_subnets(:"Subnet::Ipv4"), :first, :last,
```
Comment (tbrisker) [inline app/views/hostgroups/_form.html.erb]: `subnet.to_label != subnet.name`

File: app/views/hostgroups/_form.html.erb [hunk 11]
```diff
@@ -49,14 +49,14 @@
...
                      { :include_blank => domain_subnets(:"Subnet::Ipv4").any? ? blank_or_inherit_f(f, :subnet) : _("No subnets") },
                      { :disabled => domain_subnets(:"Subnet::Ipv4").empty?,
                        :data     => { :url     => SETTINGS[:unattended] ? freeip_subnets_path : '',
                                       :subnets => subnets_ipam_data(:"Subnet::Ipv4").to_json },
                        :label    => _("IPv4 Subnet"),
                        :help_inline => :indicator } %>
-        <%= select_f f, :subnet6_id, domain_subnets(:"Subnet::Ipv6"), :id, :to_label,
+        <%= select_f f, :subnet6_id, domain_subnets(:"Subnet::Ipv6"), :first, :last,
```
Comment (tbrisker) [inline app/views/hostgroups/_form.html.erb]: `subnet.to_label != subnet.name`

File: app/views/puppetclasses/_classes.html.erb [hunk 12]
```diff
@@ -1,16 +1,23 @@
-<% Puppetclass.classes2hash(puppetclasses).sort.in_groups(2,nil) do |group| %>
+<% authorized = authorized_for(controller: :host_editing, action: :edit_classes) %>
+
+<% puppetclasses.group_by { |klass| klass.module_name }.sort.in_groups(2, nil) do |group| %>
```
Comment (tbrisker) [inline app/views/puppetclasses/_classes.html.erb:3]: Perhaps this should be done in `Puppetclass#classes2hash` since this is the only place using that function? (or that function can be removed)

**Review by tbrisker — CHANGES_REQUESTED** [review]
A few comments inline, but in general - much of this feels like micro-optimization that reduces readability for little benefit; I'd be fine with sacrificing readability if it leads to significant improvements in performance, but many of the changes here might not meet that criteria. Perhaps there are other changes that can be done that will lead to more significant impact?

File: app/helpers/application_helper.rb (hunk 6 above)
Comment (justahero) [reply app/helpers/application_helper.rb:482]: Let me see if there is a good way to re-use existing code. This change was contributed by @timogoebel, maybe he can explain a bit more.

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (timogoebel) [reply app/helpers/application_helper.rb:488]: No, this is a rails limitation. You have to match both queries or it won't work.

**Comment by justahero** [comment]
First of all, thanks for the valuable & thorough feedback, it is really appreciated. :+1:

Most (if not all) comments are addressed now, I reverted some of the functions, while other parts were refactored to not change the code too much. As observed rightly before by @tbrisker some of the changes obfuscate / make the code harder to read while the immediate performance gains are probably negligible. For this reason I refactored the code to improve the locality while hopefully getting a few performance gains in the right places.

Instead of writing code, e.g. for a select box, in a ERB template like this

```erb
<%= select_f f, ..., accessible_resource(f.object, :architecture), :id, :name, {...}, {...} %>
```

the code changes to

```erb
<%= select_f f, ..., accessible_resource(f.object, :architecture).pluck(:id, :name), :first, :last, {...}, {...} %>
```

where appropriate. Hopefully some of the changes improve a bit of speed, while still being reasonably fair to read. I would like to run a few tests with the current change set in order to see how big the improvement of the Puppet classes view rendering is.

Once the changes are ok I would like to rebase, to make the commits clearer.

File: app/helpers/application_helper.rb [hunk 13]
```diff
@@ -475,11 +475,20 @@ def accessible_resource_records(resource, order = :name)
   end
 
   def accessible_resource(obj, resource, order = :name, association: resource)
-    list = accessible_resource_records(resource, order).to_a
```
Comment (ShimShtein) [inline app/helpers/application_helper.rb:478]: Can we cache accessible records per-request? We are reusing the same list many times inside a form.
Good example would be NICs, where for each rendered NIC, we render list of domains that shouldn't change between NICs in the same request.

I am thinking about `class => resources_list` kind of cache here, so we won't need to implement it for each field in `nics/_base_form.html.erb`.

**Review by ShimShtein — COMMENTED** [review]
From looking at profiler results, it looks like most of the time is spent in rendering NICs. Most of host records do have a lot of NICs defined, which can make the total rendering time quite big (close to 4s for me too).

From a quick glance, it looks like we could benefit a lot from on-demand loading of available resources list for our forms. I have suggested caching those lists if we repeat the same query more than once, like in the case of NICs, but maybe it would be better to do it on UI side completely, maybe @theforeman/ui-ux will have plans for refactoring select2 field as a react component?

File: app/helpers/application_helper.rb (hunk 13 above)
Comment (amirfefer) [reply app/helpers/application_helper.rb]: @ShimShtein  we do have a plan to use a react implementation for select2, in fact the switcher (from breadcrumbs switcher) is almost a  select with search ability already.

@glekner have you started to work on select2 component ?

**Comment by amirfefer** [comment]
@ShimShtein  we do have a plan to use a react implementation for select2, in fact the switcher (from breadcrumbs switcher) is almost a  select with search ability already.

@glekner have you started to work on select2 component ?

**Comment by ShimShtein** [comment]
@amirfefer make sure it is capable of both loading a static list of options and loading the options from additional API call. In this use case we are looking for the second option.

**Comment by timogoebel** [comment]
> From looking at profiler results, it looks like most of the time is spent in rendering NICs. Most of host records do have a lot of NICs defined, which can make the total rendering time quite big (close to 4s for me too).

I believe we did some "benchmarking" (checking individual requests) on a host with just a simple interface (common for VMs).

> I have suggested caching those lists if we repeat the same query more than once, like in the case of NICs

Shouldn't rails cache these queries automatically (if the queries are identical)?

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (tbrisker) [reply app/helpers/application_helper.rb:488]: wouldn't this prevent rails from caching the query results since it is a different query every time now?

File: app/helpers/application_helper.rb (hunk 1 above)
Comment (justahero) [reply app/helpers/application_helper.rb:494]: I think this method is then obsolete.

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (timogoebel) [reply app/helpers/application_helper.rb:488]: That might be true.

**Comment by justahero** [comment]
In a discussion between @timogoebel, @ShimShtein and me, we came to the conclusion to focus on the Puppet Class template rendering alone. This is to not conflate different optimizations which are not backed by a more detailed analysis.

Therefore I reverted a few more optimizations.

**Comment by tbrisker** [comment]
Thanks @justahero ! Looks like this has been WIP and inactive for a long time and some of the main performance improvements were already extracted to a separate PR. I'll go ahead and close this for now, feel free to reopen if you wish to continue working on it.
